### DISCUSSION TITLE
generate gif with multiple pngs?

### DISCUSSION THREAD

**Discussion by wade30822** [libvips/libvips#4291] [Q&A] [answered]
Does the libvips library support generating GIFs from multiple PNG images with control over parameters such as delay and disposal? i googled and read docs a lot and didn't find solution.  If it does, could you please provide a simple C demo?

THX～

**Comment by jcupitt** [comment, marked as answer]
Hi @wade30822,

Yes, try:

```
vips arrayjoin "1.png 2.png 3.png" x.gif[page-height=100] --across 1
```

Where `100` is the height of each of your PNG pages. If you're using bash, you could write eg.:

```
height=$(vipsheader -f height $1)
vips arrayjoin "$*" x.gif[page-height=$height] --across 1
```

libvips represents animated and multipage images as tall, thin strips of frames, like a strip of movie film (or a roll of toilet paper). `page-height` is the vertical height of each frame and must be a factor of the total image height. The delay and loop settings are passed in metadata. 

In python you can write:

```python
#!/usr/bin/env python3

import sys
import pyvips

# the input images
# assume these are all the same size
images = [pyvips.Image.new_from_file(filename, access="sequential")
          for filename in sys.argv[2:]]

# frame delays are in milliseconds
delay_array = [300] * len(images)

animation = pyvips.Image.arrayjoin(images, across=1)
animation.set_type(pyvips.GValue.gint_type, "page-height", images[0].height)
animation.set_type(pyvips.GValue.array_int_type, "delay", delay_array)
print(f"writing {sys.argv[1]} ...")
animation.write_to_file(sys.argv[1])
```

I made a C example here: 

https://github.com/libvips/libvips/blob/master/examples/assemble-animated.c

Maybe this should be a chapter in the docs.

**Reply by wade30822** [reply]
Wow! That's really awesome! Thank you for your help. You've saved my day big time.

**Comment by jcupitt** [comment]
.. I tried drafting a chapter for the docs about multipage image handling:

https://github.com/libvips/libvips/pull/4292

Any comments on clarity or missing material would be very welcome!

**Reply by wade30822** [reply]
Hi @jcupitt

There are two compilation errors with code `g_autoptr(VipsImage) strip;`
`
 error: unknown type name 'VipsImage_autoptr'
   29 |     g_autoptr(VipsImage) strip;
      |               ^~~~~~~~~
error: cleanup argument not a function
   29 |     g_autoptr(VipsImage) strip;
`
should it be 
`
    g_autoptr(GObject) strip = G_OBJECT(vips_image_new());
    VipsImage *anim = VIPS_IMAGE(strip);
`
?

**Reply by jcupitt** [reply]
autoptr support is a new thing in 8.16 (the current stable version). If you're using an older libvips, add:

```C
G_DEFINE_AUTOPTR_CLEANUP_FUNC(VipsImage, g_object_unref)
```

just after including `vips.h`.

**Reply by wade30822** [reply]
> autoptr support is a new thing in 8.16 (the current stable version). If you're using an older libvips, add:
> 
> ```c
> G_DEFINE_AUTOPTR_CLEANUP_FUNC(VipsImage, g_object_unref)
> ```
> 
> just after including `vips.h`.

Got it! Thx a lot.
